### ISSUE TITLE
Are priors included in the logLik output?

### ISSUE THREAD

**Issue by paul-buerkner** [glmmTMB/glmmTMB#1116]
When I add priors to glmmTMB, I sometimes can unexpected logLik outputs, especially when the model has many parameters. An inspection of `logLik.glmmTMB` seems to indicate that it uses the `fit$objective` element to compute the log-likelihood. I assume the `objective` also includes the log prior contributions(?) If that is the case, I believe the logLik output is incorrect for models with priors. Would there be a way to compute the log-likelihood without including the log prior contribution?

**Comment by bbolker** [comment]
Yes, this also includes the prior contributions.  Here is (I think) a hack-y way to get the log-likelihood:

```
library(glmmTMB)
data("sleepstudy", package = "lme4")
prior1 <- data.frame(prior = c("normal(250,3)","t(0,3,3)","gamma(10,1)"),
                     class = c("fixef", "fixef", "ranef_sd"),
                     coef = c("(Intercept)", "Days", "Subject"))
## fit initial model, with and without (for comparison) priors
g0 <- glmmTMB(Reaction ~ 1 + Days + (1 + Days |Subject), sleepstudy)
g1 <- update(g0, priors = prior1)

## now hack ...

## extract fitted parameters (don't include latent variables. We might have to work a little harder
##  if we had random effects in the zero-inflation or dispersion models
pars <- g1$obj$env$parList()
pars2 <- pars[lengths(pars)>0 & names(pars)!="b"]
## set things up to fix these parameters to starting values
mapvec <- lapply(pars2, \(x) factor(rep(NA_real_, length(x))))

g2 <- update(g1, priors = NULL, start = pars2, map = mapvec)

## compare:
logLik(g0)
logLik(g1)
logLik(g2)
```

These answers seem reasonable but I haven't checked them carefully.

Deeper question, how do we support doing this in a non-hacky way?  Add an "include prior in objective function computation" flag to the TMB code that we can toggle?

**Comment by paul-buerkner** [comment]
Thank you a lot! We will take a look at see if this indeed also fixes it for your cases.

> Deeper question, how do we support doing this in a non-hacky way? Add an "include prior in objective function computation" flag to the TMB code that we can toggle?

I have very limited experience with glmmTMB but something like this seems like a good solution to me.

**Comment by paul-buerkner** [comment]
In our experiments, we had set `glmmTMBControl(profile = TRUE)`, in which case the above hack fails because `Q` in `fitTMB` is NULL. Is it safe to deactivate via `profile = FALSE` in the hack call to `update`?

**Comment by bbolker** [comment]
I think it should be ...

**Comment by jennyp76** [comment]
@bbolker
Do you have any updates on this suggestion??

> Deeper question, how do we support doing this in a non-hacky way? Add an "include prior in objective function computation" flag to the TMB code that we can toggle?

**Comment by bbolker** [comment]
Can you say a little bit more about your use case? Why are you interested in recovering the value of the log-likelihood (without the influence of priors)?<sup>*</sup> I agree that listing the maximum unnormalized log posterior density (ugh) as "log-likelihood" is incorrect; from a narrow technical sense, we could solve this by changing the label in the output (and having 'logLik' set to NA, or issuing a warning).

<sup>*</sup>I'm not saying you shouldn't want to do this, just wondering about how you use this information in your modeling/inferential workflow.

**Comment by jennyp76** [comment]
I have very limited experience with glmmTMB therefore I’m not sure if I am on the same context as you. Hope I'm right :) 

My use case is a beta-binomial GLMM for allele-specific methylation counts.

The full model is

    cbind(methylated, unmethylated) ~ haplotype + (1 | individual)

and the reduced model is

    cbind(methylated, unmethylated) ~ 1 + (1 | individual)

Both models use `betabinomial(link = "logit")` with `dispformula = ~1`. I apply the same normal prior to the dispersion intercept, i.e. log(phi), in both models.

After fitting the models, I would like to evaluate how well the fitted MAP parameter values explain the observed data alone, separately from how well the estimated dispersion agrees with the prior. I want the data-only marginal log-likelihood evaluated at the MAP estimates.

For the current case, where there is only a single normal prior on the dispersion intercept, I am planning to calculate

    `logLik_data_at_MAP = as.numeric(logLik(fit)) - dnorm(fixef(fit)$disp["(Intercept)"], mean = prior_mean, sd = prior_sd, log = TRUE)`

However, an official interface such as `logLik(fit, include_prior = FALSE)` or separate accessors for the data log-likelihood and the penalized
log-posterior objective would be safer and more general than relying on internal objects or a manual correction.

Would this be a reasonable use case for the proposed flag? Could you also confirm whether the subtraction above correctly recovers the marginal data log-likelihood evaluated at the MAP estimates, particularly for a non-Gaussian mixed model using the Laplace approximation?

Thank you
Jen
